README.md
# sgamesolver  
  
**Stochastic game solver**, short: sgamesolver, is a python package to compute 
stationary or Markov-perfect equilibria of stochastic games, using the 
homotopy continuation method.

Some useful links:
- Source code is hosted on [github](https://github.com/davidpoensgen/sgamesolver)
- Also on github is an [issue tracker](https://github.com/davidpoensgen/sgamesolver/issues) for bug reports and feature requests
- Documentation is found on [readthedocs](https://sgamesolver.readthedocs.io/en/latest/)

sgamesolver is free and open source, under the MIT license. 
If you use the program for any published research, please cite 
[Eibelshäuser/Poensgen (2019)](https://dx.doi.org/10.2139/ssrn.3316631).

## Installation
sgamesolver is hosted on PyPI, so installation is usually as simple as
 ````sh
pip install sgamesolver
````
Head over to the docs for other options for installation.

## Usage
Solving a stochastic game is done in three steps: 
- define a game
- pick a homotopy function
- set up and run the solver

````python
import sgamesolver

game = sgamesolver.SGame.random_game(64, 2, 4, seed=42)

qre = sgamesolver.homotopy.QRE(game)

qre.solver_setup()
qre.solve()

print(qre.equilibrium)
````


A quick rundown of these steps – more details are found in the docs:

##### 1. Set up a stochastic game
````python
game = sgamesolver.SGame.random_game(64, 2, 4, seed=42)
````

Stochastic games are represented by the class `SGame`. 
For this quick example, we are using the method `random_game` to 
randomize a game with 64 states, 2 players, and 4 actions per player 
and state. (Setting a seed just makes the result reproducible.) 

Of course, you probably didn't come here to solve a random game, 
but have a specific game in mind. The 
[documentation](https://sgamesolver.readthedocs.io/en/latest/) contains instructions 
and examples on how to create an `SGame` which represents the stochastic 
game you want to solve. (By the way, sgamesolver can also solve normal form games, 
sequential games, repeated games, and also 
finite Markov decision problems – each of these is just a simple case of a stochastic game.)  


##### 2. Select and set up a homotopy function for your game
```python
qre = sgamesolver.homotopy.QRE(game)
```

sgamesolver uses the **homotopy principle** to solve stochastic games, 
a general technique to solve systems of non-linear equations. 
In short, the idea is as follows: Instead of solving some very hard problem directly
(in our case: finding an equilibrium), a continuous transformation 
is applied to the system, yielding a related, but much simpler problem, 
for which one can easily obtain a solution. This transformation is then 
gradually reversed while tracking the solution, until arriving at a solution 
for the original problem – here, the desired stationary equilibrium. 
(You can find more background in the documentation – 
although such knowledge is not necessary for using the program.)

The (mathematical) function used for this transformation is called 
**homotopy function**. In general, there are many possibilities 
to construct a suitable one. sgamesolver currently includes two: 
The one we picked for this example, `sgamesolver.homotopy.QRE`, is based on 
an extension of quantal response equilibrium to stochastic games 
 [(Eibelshäuser/Poensgen 2019b)](https://dx.doi.org/10.2139/ssrn.3314404). The other, `sgamesolver.homotopy.LogTracing`, implements the 
logarithmic tracing procedure for stochastic games [(Eibelshäuser/Klockmann/Poensgen/von Schenk 2022)](https://dx.doi.org/10.2139/ssrn.3748830). 
Which one to pick? In our experience, the former is more robust – 
while the latter has the advantage that it allows to search for multiple equilibria. 
More homotopy functions are to come! In any case, please make sure to cite the paper that 
introduced whichever homotopy you end up using.

##### 3. Let the homotopy solver do its job
Finally, we will set up the solver and start it:
```python
qre.solver_setup()
qre.solve()
```

Then it's time to lean back and watch for a bit:
```
==================================================
Start homotopy continuation
Step    37: t =  3.612 ↑, s =  20.47, ds =  3.418
```
... until ...
```
Step   247: t = 1.147e+04 ↑, s = 9.385e+04, ds =  1000.
Step   247: Continuation successful. Total time elapsed: 0:00:15
End homotopy continuation
==================================================
An equilibrium was found via homotopy continuation.
```
... success!
Ideally, the solver will be able to find a solution without
requiring any further interaction, as in this example. In cases 
where this does not work out, check out the 
[section on troubleshooting](https://sgamesolver.readthedocs.io/en/latest/troubleshooting.html) 
in the documentation.

##### 4. Aftermath
We can now display the solution:
```python
print(qre.equilibrium)
```
which outputs equilibrium strategies and values for all 64 states:
```
+++++++++ state00 +++++++++
                       a0    a1    a2    a3  
player0 : v=15.09, σ=[1.000 0.000 0.000 0.000]
player1 : v=15.63, σ=[0.000 0.000 1.000 0.000]

+++++++++ state01 +++++++++
                       a0    a1    a2    a3  
player0 : v=14.76, σ=[0.000 0.961 0.000 0.039]
player1 : v=15.61, σ=[0.354 0.000 0.000 0.646]

+++++++++ state02 +++++++++
                       a0    a1    a2    a3  
player0 : v=14.84, σ=[1.000 0.000 0.000 0.000]
player1 : v=15.61, σ=[0.000 0.000 1.000 0.000]

... (abridged here for brevity) ...

+++++++++ state63 +++++++++
                       a0    a1    a2    a3  
player0 : v=14.92, σ=[0.000 1.000 0.000 0.000]
player1 : v=15.75, σ=[1.000 0.000 0.000 0.000]
```
Of course, you now also can access equilibrium strategies (and values) 
as `numpy` arrays and use them for further calculations or simulations.
```
eq_strat = qre.equilibrium.strategies
eq_values = qre.equilibrium.values
```




CHANGELOG.md
# Change Log

## [1.0.2] - 2023-08-03

### New
- Added basic functionality for simulations. Implemented as method `.simulate()` 
  of the class `StrategyProfile`, e.g. a computed equilibrium. 
- Added the option to create non-generic random games. (`SGame.random_nongeneric_game`).
  Can be used with the `sgamesolver-timings` terminal script by specifying "nongeneric=True" as 
  game parameter in the Excel sheet. 

### Changed
- Parameter 'eta_fix' of Logtracing now defaults to True.

### Fixed
- Fixed some bugs regarding error reporting in `SGame.from_table()`
  (which occurred when the action labels where not natively strings).

## [1.0.1] - 2022-09-04

### New
- `SGame.from_table()` is more flexible now: to_state-column now also accepts a strings 
  formatted as 'state: prob, state: prob, ...'. See documentation for details.

## [1.0.0] - 2022-06-06

### New
- Everything!


dsgamesolver/ipm.py
"""Interior point method homotopy."""

# TODO: check user-provided initial_strategies and weights?
# TODO: adjust tracking parameters with "scale" of game
# TODO: think about Cython import
# TODO: add Numpy implementation?


from typing import Union, Optional

import numpy as np
from numpy.typing import ArrayLike

from dsgamesolver.sgame import SGame, SGameHomotopy, LogStratHomotopy
from dsgamesolver.homcont import HomCont


ABC = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'


# %% parent class for interior point method homotopy


class IPM(SGameHomotopy):
    """Interior point method homotopy: base class"""

    def __init__(self, game: SGame, initial_strategies: Union[str, ArrayLike] = "centroid",
                 weights: Optional[ArrayLike] = None) -> None:
        super().__init__(game)

        # TODO: adjust parameters with scale of payoff matrix:

        self.tracking_parameters['normal'] = {
            'x_tol': 1e-7,
            't_tol': 1e-7,
            'H_tol': 1e-7,
            'ds0': 0.1,
            'ds_infl': 1.2,
            'ds_defl': 0.5,
            'ds_min': 1e-9,
            'ds_max': 1000,
            'corr_steps_max': 20,
            'corr_dist_max': 0.5,
            'corr_ratio_max': 0.9,
            'detJ_change_max': 1.5,
            'bifurc_angle_min': 175,
        }

        self.tracking_parameters['robust'] = {
            'x_tol': 1e-7,
            't_tol': 1e-7,
            'H_tol': 1e-8,
            'ds0': 0.1,
            'ds_infl': 1.1,
            'ds_defl': 0.5,
            'ds_min': 1e-9,
            'ds_max': 10,
            'corr_steps_max': 30,
            'corr_dist_max': 0.3,
            'corr_ratio_max': 0.7,
            'detJ_change_max': 1.3,
            'bifurc_angle_min': 175,
        }

        if initial_strategies == "centroid":
            self.sigma_0 = np.zeros((self.game.num_states, self.game.num_players, self.game.num_actions_max))
            for s in range(self.game.num_states):
                for p in range(self.game.num_players):
                    self.sigma_0[s, p, 0:self.game.nums_actions[s, p]] = 1 / self.game.nums_actions[s, p]
        elif initial_strategies == "random":
            self.sigma_0 = np.zeros((self.game.num_states, self.game.num_players, self.game.num_actions_max))
            for s in range(self.game.num_states):
                for p in range(self.game.num_players):
                    sigma = np.random.exponential(scale=1, size=self.game.nums_actions[s, p])
                    self.sigma_0[s, p, 0:self.game.nums_actions[s, p]] = sigma / sigma.sum()
        else:
            # TODO: document how initial strategies should be specified / should they be checked?
            self.sigma_0 = np.array(initial_strategies)

        self.sigma_0_flat = self.game.flatten_strategies(self.sigma_0)

        if weights is None:
            self.nu = np.ones((self.game.num_states, self.game.num_players, self.game.num_actions_max))
        else:
            # TODO: document how weights should be specified / should they be checked?
            self.nu = weights

    def initialize(self) -> None:
        self.y0 = self.find_y0()
        # Note: homotopy parameter t goes from 1 to 0
        self.solver = HomCont(self.H, self.y0, self.J, t_target=0.0,
                              parameters=self.tracking_parameters['normal'],
                              x_transformer=self.x_transformer, store_path=True)

    def find_y0(self) -> np.ndarray:
        V = np.ones((self.game.num_states, self.game.num_players))
        return self.sigma_V_t_to_y(self.sigma_0, V, 1.0)

    def x_transformer(self, y: np.ndarray) -> np.ndarray:
        x = y.copy()
        z, t = x[0:self.game.num_actions_total], x[-1]
        x[0:self.game.num_actions_total] = 0.25 * (z + (z**2 + 4*t*self.sigma_0_flat**0.5)**0.5)**2
        return x

    def sigma_V_t_to_y(self, sigma: np.ndarray, V: np.ndarray, t: float) -> np.ndarray:
        sigma_flat = self.game.flatten_strategies(sigma)
        z_flat = (sigma_flat - t * self.sigma_0_flat**0.5) / sigma_flat**0.5
        V_flat = self.game.flatten_values(V)
        return np.concatenate([z_flat, V_flat, [t]])

    def y_to_sigma_V_t(self, y: np.ndarray, zeros: bool = False) -> tuple[np.ndarray, np.ndarray, float]:
        sigma_V_t_flat = self.x_transformer(y)
        sigma = self.game.unflatten_strategies(sigma_V_t_flat[0:self.game.num_actions_total], zeros=zeros)
        V = self.game.unflatten_values(sigma_V_t_flat[self.game.num_actions_total:-1])
        t = sigma_V_t_flat[-1]
        return sigma, V, t


# %% Cython implementation of IPM


class IPM_ct(IPM):
    """Interior point method homotopy: Cython implementation"""

    def __init__(self, game: SGame, initial_strategies: Union[str, ArrayLike] = "centroid",
                 weights: Optional[ArrayLike] = None) -> None:
        super().__init__(game, initial_strategies, weights)

        # only import Cython module on class instantiation
        try:
            import pyximport
            pyximport.install(build_dir='./dsgamesolver/__build__/', build_in_temp=False, language_level=3,
                              setup_args={'include_dirs': [np.get_include()]})
            import dsgamesolver.ipm_ct as ipm_ct

        except ImportError:
            raise ImportError("Cython implementation of IPM homotopy could not be imported. ",
                              "Make sure your system has the relevant C compilers installed. ",
                              "For Windows, check https://wiki.python.org/moin/WindowsCompilers ",
                              "to find the right Microsoft Visual C++ compiler for your Python version. ",
                              "Standalone compilers are sufficient, there is no need to install Visual Studio. ",
                              "For Linux, make sure the Python package gxx_linux-64 is installed in your environment.")

        self.ipm_ct = ipm_ct

    def H(self, y: np.ndarray) -> np.ndarray:
        return self.ipm_ct.H(y, self.game.payoffs, self.game.transitions, self.sigma_0, self.nu,
                             self.game.num_states, self.game.num_players, self.game.nums_actions,
                             self.game.num_actions_max, self.game.num_actions_total)

    def J(self, y: np.ndarray) -> np.ndarray:
        return self.ipm_ct.J(y, self.game.payoffs, self.game.transitions, self.sigma_0, self.nu,
                             self.game.num_states, self.game.num_players, self.game.nums_actions,
                             self.game.num_actions_max, self.game.num_actions_total)


# %% Sympy implementation of IPM


class IPM_sp(IPM):
    """Interior point method homotopy: Sympy implementation"""

    def __init__(self, game: SGame, initial_strategies: Union[str, ArrayLike] = "centroid",
                 weights: Optional[ArrayLike] = None) -> None:
        super().__init__(game, initial_strategies, weights)

        # only import Sympy module on class instantiation
        try:
            import sympy as sp

        except ModuleNotFoundError:
            raise ModuleNotFoundError("Sympy implementation of IPM homotopy requires package 'sympy'.")

        sp.init_printing()

        num_s, num_p, nums_a = self.game.num_states, self.game.num_players, self.game.nums_actions
        num_a_max, num_a_tot = self.game.num_actions_max, self.game.num_actions_total

        # symbols
        y = sp.symarray('y', num_a_tot + num_s*num_p + 1)

        # strategies, values and homotopy parameter

        z = np.zeros((num_s, num_p, num_a_max), dtype='object')
        flat_index = 0
        for s in range(num_s):
            for p in range(num_p):
                for a in range(nums_a[s, p]):
                    z[s, p, a] = y[flat_index]
                    flat_index += 1

        V = np.zeros((num_s, num_p), dtype='object')
        flat_index = num_a_tot
        for s in range(num_s):
            for p in range(num_p):
                V[s, p] = y[flat_index]
                flat_index += 1

        t = y[-1]

        # transformations
        sigma = 0.25*(z + (z**2 + 4*t*self.sigma_0**0.5)**0.5)**2
        lambda_ = 0.25*(-z + (z**2 + 4*t*self.sigma_0**0.5)**0.5)**2

        # payoffs including continuation values

        # u_tilde[spA]
        u_tilde = np.zeros(self.game.payoffs.shape, dtype='object')
        u_tilde += self.game.payoffs
        for index, _ in np.ndenumerate(self.game.payoffs):
            for to_state in range(num_s):
                u_tilde[index] += (self.game.transitions[(*index, to_state)] * V[to_state]).sum()

        # Eu_tilde_a[spa]
        Eu_tilde_a = np.zeros((num_s, num_p, num_a_max), dtype='object')
        for index, util in np.ndenumerate(u_tilde):
            temp_prob = 1
            for other in range(num_p):
                if other == index[1]:
                    continue
                temp_prob *= sigma[index[0], other, index[other+2]]
            Eu_tilde_a[index[0], index[1], index[index[1]+2]] += temp_prob * util

        # homotopy function
        H_list = []
        for s in range(num_s):
            for p in range(num_p):
                for a in range(nums_a[s, p]):
                    H_list.append((1-t)*Eu_tilde_a[s, p, a] + lambda_[s, p, a] - V[s, p] - t*(1-t)*self.nu[s, p, a])
        for s in range(num_s):
            for p in range(num_p):
                H_list.append(sigma[s, p, 0:nums_a[s, p]].sum() - 1)
        H_sym = sp.Matrix(H_list)

        # Jacobian matrix
        J_sym = H_sym.jacobian(y)

        # lambdify
        self.H_num = sp.lambdify(y, H_sym, modules=['numpy'])
        self.J_num = sp.lambdify(y, J_sym, modules=['numpy'])

    def H(self, y: np.ndarray) -> np.ndarray:
        return self.H_num(*tuple(np.array(y)))[:, 0]

    def J(self, y: np.ndarray) -> np.ndarray:
        return self.J_num(*tuple(np.array(y)))


# %% testing


if __name__ == '__main__':

    from tests.random_game import create_random_game
    game = SGame(*create_random_game())

    # cython
    ipm_ct = IPM_ct(game)
    ipm_ct.initialize()
    ipm_ct.solver.solve()

    y0 = ipm_ct.y0
    """
    %timeit ipm_ct.H(y0)
    %timeit ipm_ct.J(y0)
    """

    # sympy
    ipm_sp = IPM_sp(game)
    ipm_sp.initialize()
    ipm_sp.solver.verbose = 2
    ipm_sp.solver.solve()

    """
    assert np.allclose(ipm_sp.y0, y0)
    %timeit ipm_sp.H(y0)
    %timeit ipm_sp.J(y0)
    """
